src/Magnum/DartIntegration/Object.cpp

```diff
@@ -0,0 +1,519 @@
+
+    THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+    IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+    FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL
+    THE AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+    LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING
+    FROM, OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER
+    DEALINGS IN THE SOFTWARE.
+*/
+
+#include "Object.h"
+
+#include <dart/dynamics/BodyNode.hpp>
+#include <dart/dynamics/BoxShape.hpp>
+#include <dart/dynamics/CapsuleShape.hpp>
+#include <dart/dynamics/CylinderShape.hpp>
+#include <dart/dynamics/EllipsoidShape.hpp>
+#include <dart/dynamics/MeshShape.hpp>
+#include <dart/dynamics/ShapeNode.hpp>
+#include <dart/dynamics/SoftBodyNode.hpp>
+#include <dart/dynamics/SoftMeshShape.hpp>
+#include <dart/dynamics/SphereShape.hpp>
+
+#include <Corrade/PluginManager/Manager.h>
+#include <Corrade/Utility/Directory.h>
+
+#include <Magnum/Buffer.h>
+#include <Magnum/Mesh.h>
+#include <Magnum/PixelFormat.h>
+#include <Magnum/Texture.h>
+#include <Magnum/TextureFormat.h>
+#include <Magnum/MeshTools/Compile.h>
+#include <Magnum/MeshTools/Transform.h>
+#include <Magnum/Primitives/Capsule.h>
+#include <Magnum/Primitives/Cube.h>
+#include <Magnum/Primitives/Cylinder.h>
+#include <Magnum/Primitives/Icosphere.h>
+#include <Magnum/Trade/AbstractImporter.h>
+#include <Magnum/Trade/ImageData.h>
+#include <Magnum/Trade/MeshData3D.h>
+#include <Magnum/Trade/MeshObjectData3D.h>
+#include <Magnum/Trade/ObjectData3D.h>
+#include <Magnum/Trade/PhongMaterialData.h>
+#include <Magnum/Trade/TextureData.h>
+
+static Eigen::Vector3d normalFromVertex(const dart::dynamics::SoftBodyNode* bn,
+                                        const Eigen::Vector3i& face,
+                                        std::size_t v)
+{
+  const Eigen::Vector3d& v0 = bn->getPointMass(face[v])->getLocalPosition();
+  const Eigen::Vector3d& v1 = bn->getPointMass(face[(v+1)%3])->getLocalPosition();
+  const Eigen::Vector3d& v2 = bn->getPointMass(face[(v+2)%3])->getLocalPosition();
+
+  const Eigen::Vector3d dv1 = v1-v0;
+  const Eigen::Vector3d dv2 = v2-v0;
+  const Eigen::Vector3d n = dv1.cross(dv2);
+
+  double weight = n.norm()/(dv1.norm()*dv2.norm());
+  weight = std::max( -1.0, std::min( 1.0, weight) );
+
+  return n.normalized() * asin(weight);
+}
+
+static void computeNormals(std::vector<Eigen::Vector3d>& normals,
+                           const dart::dynamics::SoftBodyNode* bn)
+{
+  for(std::size_t i=0; i<normals.size(); ++i)
+    normals[i] = Eigen::Vector3d::Zero();
+
+  for(std::size_t i=0; i<bn->getNumFaces(); ++i)
+  {
+    const Eigen::Vector3i& face = bn->getFace(i);
+    for(std::size_t j=0; j<3; ++j)
+      normals[face[j]] += normalFromVertex(bn, face, j);
+  }
+
+  for(std::size_t i=0; i<normals.size(); ++i)
+    normals[i].normalize();
+}
```

Comment:
> This whole thing could be replaced by MeshTools::generateFlatNormals(), I think. Correct me if I'm wrong, of course ;)

Initially I was using the `MeshTools::generateFlatNormals()` functionality, but then I realized that their SoftMeshes are ''special''. Long story short, the triangles generated in the mesh do not all follow the same orientation (i.e., either clockwise or anti-clockwise) and as such I found this code in their own GUI implementations in order to try to fix this issue. The result is still not the best:

![image](https://user-images.githubusercontent.com/1283922/36358440-b0d066d8-150e-11e8-8eb4-2443badf6c5f.png)

Nevertheless, it is much better than generating the normals assuming that they follow some orientation; in that case, we get wrong lighting (e.g., triangles completely black etc)..

As a side note, I found this [nice library](https://github.com/libigl/libigl) that has a nice [facet orientation fixing algorithm](http://libigl.github.io/libigl/tutorial/tutorial.html#facetorientation). I might  at some point re-implement their algorithm for Magnum (it might be a nice feature) and thus we could use it here to get much nicer normals..

Reply:
Right, sorry, I remember we were talking about this.

I looked at the algorithm only briefly at first and thought it's generating flat normals (didn't see the `+=` at the end). Now I see that it's generating smooth normals: averaging the normal over neighboring faces, doing *some* weighting on top (I'm not sure if completely correctly, as I see some weird artifacts on the screenshot). That fixes the face winding issue only by accident, I think -- only if there's not many faces with wrong winding.

Compared to the flat normals the result is quite ugly, though :/ 

Most importantly, why I'm commenting: this code snippet is never going to be tested well enough in the context of the DartIntegration library and since it doesn't really solve the problem in a nice way, I don't know if it's worth to invest further time in this direction (I originally wanted to suggest extracting this code out into MeshTools and test it there).

But I got two different ideas:

1. what if you would, in case of those "special SmoothMeshes", add every face twice, once in the original winding and once in the reverse? That, together with enabling face culling and generating flat normals afterwards, could solve the issue in a nice-looking (yet ugly) way. This would need to be explicitly mentioned in the docs, as forgetting to enable face culling may have disastrous results, heh :)
2. what if you would use functionality from Assimp itself to generate the normals? That would remove the need for the above code, avoid extra testing effort and we could blame a third-party lib for producing ugly artefacts :laughing: The [relevant option is here](http://assimp.sourceforge.net/lib_html/postprocess_8h.html#a64795260b95f5a4b3f3dc1be4f52e410a6afb4fee42eca4482674859196cb8685). But I give no guarantees that this approach would also *accidentally circumvent* the face winding issue (and it will still look ugly). So maybe solution 1 is still the best among the worst :D

By the way, did you report this "speciality" to DART? Fixing it there would be the best solution in my opinion.

Reply:
I did a bit of research, opened https://github.com/mosra/magnum/pull/229 so I don't forget all the details

Reply:
> what if you would, in case of those "special SmoothMeshes", add every face twice, once in the original winding and once in the reverse? That, together with enabling face culling and generating flat normals afterwards, could solve the issue in a nice-looking (yet ugly) way. This would need to be explicitly mentioned in the docs, as forgetting to enable face culling may have disastrous results, heh :)

I think I like this idea more..! Let me finish the `convertShapeNode()` first and then I'll do this.. Thanks for the suggestions..

Reply:
> By the way, did you report this "speciality" to DART? Fixing it there would be the best solution in my opinion.

They already know it! :stuck_out_tongue: I can make a more explicit issue with possible solutions (or even fix it).. Putting it on my TO-DO list..!

Reply:
I tried to search, but didn't find any issue related to this on their main repo.

Adding a link to the issue to the documentation about "why is everything twice and why one needs to enable face culling" would be great, so somebody can look it up later and discard the whole workaround once they fix it :)

Reply:
> Adding a link to the issue to the documentation about "why is everything twice and why one needs to enable face culling" would be great, so somebody can look it up later and discard the whole workaround once they fix it :)

The thing is that they do not really care about the graphics part of the library. They have it (and update it) only when they need something (or someone does it for them). That's why I said it's better if I implement it myself for them.. :stuck_out_tongue: